Dockerfile

```diff
@@ -0,0 +1,9 @@
+FROM elixir:1.3.4
+RUN mkdir -p /app
+WORKDIR /app
+CMD ["iex", "-S", "mix"]
+COPY . /app
+RUN mix local.hex --force && \
+    mix deps.get && \
+    mix compile && \
+    mix credo --strict
```

Comment:
Sure.  "Bad code" may have been a strong phrase to use, but there are definitely situations where following credo _strictly_ will lead to forced changes to good code.

One of credo's rules is "Pipe chain should start with a raw value" but there are lots of functions that take two arguments that make a lot more sense  when those arguments are together and do not make much sense when the first argument is on a line by itself.  Suppose we have a `Matrix.new/2` function that takes number of rows and number of columns as input.  We might write

```
Matrix.new(2, 2)
|> Matrix.set(0, 0, 1)
|> Matrix.set(1, 1, 1)
```
but credo will insist that we should write
```
2
|> Matrix.new(2)
|> ...
```
or we need to introduce another variable
```
m = Matrix.new(2,2)
m
|> ...
```
Those aren't _terrible_ solutions, but there was really nothing wrong with the first one and we would be forced to make a change not for any practical reason but because we need to satisfy credo.

Another one that I personally run into frequently is "Nested modules could be aliased at the top of the invoking module".  This is generally fantastic advice because it makes refactoring significantly easier.  However, it's not unusual to run into name conflicts.  Take Ecto.Repo and MyApp.Repo for example.  Or MyApp.Query.User and MyApp.User.

Another one that bugs me is "Large numbers should be written with underscores".  That just doesn't always make sense because numbers don't always represent cardinality.  Suppose I need to construct an 8-bit binary mask.. credo will force me to write `11_111_111`.  Those underscores are supposed to be stand-ins for commas.  Binary masks don't have commas.  Another, admittedly more debatable, case is "magic" database ids - `@special_foo 12_345` - we don't do things like `SELECT * FROM foos WHERE id = 12,345` so it's weird to me to write that number with underscores.*

I'm not suggesting that it's good to _ignore_ credo.  I use it myself and I put a fair amount of effort into making my code pass as many checks as possible.  What I'm really concerned about here is having the build be strictly contingent upon it.  There are lots of examples where one could write perfectly reasonable code that causes credo to issue warnings.  In the most optimistic cases there are reasonable alternative solutions that would resolve the credo warnings, but the problem is you've changed from one reasonable solution to another _just to make credo happy_.  In some cases you're forced to choose less-clear code.

I would absolutely 100% have credo be part of the project.  You could run credo as part of the build so it spits out the output.  You could wire it up to some kind of tool that comments on pull requests. 

* Rest assured that `@special_foo 12345` is a pain inflicted upon us by some other developer.